ccmlib/common.py

```diff
@@ -76,6 +77,36 @@ class UnavailableSocketError(CCMError):
     pass
 
 
+class LogPatternToVersion(object):
+
+    def __init__(self, versions_to_patterns, default_pattern):
+        self.versions_to_patterns = versions_to_patterns
+        if "0" not in self.versions_to_patterns:
```

Comment:
I'm not sure I follow how the existing implementation precludes `default` being `None`, but I think you're on the right track. I think something like

```python
def __init__(self, versions_to_patterns, default=None):
    self.version_to_patterns, self.default = versions_to_patterns, default

def __call__(self, version):
    keys_less_than_version = [k for k in self.versions_to_patterns if k <= version]
    
    if not keys_less_than_version:
        if self.default is not None:
            return 
        raise ValueError(...)
    return self.versions_to_patterns[max(keys_less_than_version, key=...)
```

would do the trick and be readable.